Initial support for running with android emulator

jgraham opened the pull request:
This enables `wpt run firefox-android <path>` in the x86_64 emulator. Most of the settings are taken from gecko, so it won't yet generalise to other products (unless they want to use the same configuration), but it's a start to get our android support into a better shape.

jgraham committed:
Add tooltool to tools/third_party

This is used by mozilla to download static artifacts required in automation.
We want it here to download the NVD files for android tests.

jgraham committed:
Add support for installing the Android x86_64 emulator

This is a precusrsor to allowing local testing in emulated andrid
devices. It adds install-android-emulator and start-android-emulator
commands that ensure that the emulator is present and start it,
respectively.

The settings are so far just copied from what gecko does; it's assumed
that much of this will need to be configurable in the future.

jgraham committed:
Add support for running firefox_android in a local emulator

This is very basic support for running Firefox tests in an
emulator. The actual app used is by default the geckoview test apk
which matches what gecko does in CI.

Because the startup sequence for android is a little different from
desktop browsers the division of labour in run.py between install and
setup_kwargs is a little wonky; we can't install until a device is
connected or the emulator is started, and knowing that requires
setting the kwargs. So the initial cut at this just makes install()
record some state that we inspect in setup_kwargs to indicate whether
we should go ahead and run the install.

The actual apk we install is just the latest build from
mozilla-central which is pretty close to "latest nightly". It might be
possible to do better here and get the actual latest nightly.

Hexcles commented on tools/wpt/android.py:
```diff
@@ -0,0 +1,166 @@
+        raise NotImplementedError
+
+    os_name = 'darwin' if os_name == 'macosx' else os_name
+    url = 'https://dl.google.com/android/repository/sdk-tools-%s-4333796.zip' % (os_name,)
```
Is there a fixed path for the latest SDK? If we want to pin the SDK version, I'd prefer to do that in a manifest/config file.

jgraham replied:
I just copied https://searchfox.org/mozilla-central/source/python/mozboot/mozboot/android.py#186 I assume that's pinned to something and that pinning has a reason, but I don't know what it is. I agree that we should make this more generic somehow, but idk how advantageous a config file is. What's the use case you have in mind?

Hexcles replied:
Not at the moment, but based on my limited Android dev experience, SDKs are generally backward compatible (and you can select a wide range of AVD versions from the AVD manager despite the version of SDK itself), so I'd hope the latest version would just work.

Anyway, it doesn't matter too much at the moment; I'd be happy to see a TODO here and leave this as is.

Hexcles commented on tools/wpt/browser.py:
```diff
@@ -416,7 +420,33 @@ class FirefoxAndroid(Browser):
+            dest = os.pwd
+
+        TC_QUEUE_BASE = "https://queue.taskcluster.net/v1/"
+        TC_INDEX_BASE = "https://index.taskcluster.net/v1/"
```
FYI this will be the first dependency on Taskcluster domain in this repo.

jgraham replied:
Yeah, it's not ideal. If I find somewhere else that this is mirrored to I'll use that instead.

Hexcles commented on tools/wpt/mach-emulator.manifest:
```diff
@@ -0,0 +1,10 @@
+[
```
Also, I assume this would only work on Linux? And does the host have to 32-bit?

jgraham replied:
I actually don't know anything about this either. I just know that these are the NVDs that gecko uses in automation, but I haven't got as far as digging into the content to work out why they're special.

mjzffr commented on tools/wpt/android.py:
```diff
@@ -0,0 +1,166 @@
+    if not os.path.exists(sdk_manager_path):
+        raise OSError("Can't find sdkmanager at %s" % sdk_manager_path)
+
+    #TODO: Not sure what's really needed here
```
Looks right. https://hg.mozilla.org/mozilla-central/file/tip/python/mozboot/mozboot/android-packages.txt has newer versions.

I assume you're generally following https://hg.mozilla.org/mozilla-central/file/tip/python/mozboot/mozboot/android.py

jgraham replied:
Yeah, mostly following that.

jgraham committed:
Docker image updates for running android emulator

jgraham committed:
Update docker image

jgraham committed:
Skip wptrunner setup test for firefox_android

There isn't a way to make this work on CI until we can run
in priviledged mode or similar, giving kvm access. Just
skip for now.

jgraham commented:
So this is enough for me to run it locally. I've had to disable the tests in CI because in docker it wants access to kvm, which requires `--priviledged` and isn't possible on TaskCluster (using the workers available to us at present). Locally under docker I also had some problems since running with `-gpu on` failed due to the lack of GPU and `-gpu off` caused the test app to crash on startup.

I know the current state of this isn't perfect, but I'm inclined to request that we take the attitude that we'll fix some of the issues as we get CI support and support for other browsers; I don't want to spend too much time predicting what's required there since we'll find out soon enough.

mjzffr reviewed (approved):
I'm not exactly an Android automation expert, but the bits I'm familiar with look reasonable (i.e. they look similar enough to the tools I've hacked on in mozilla-central)

Hexcles commented on tools/docker/Dockerfile:
```diff
@@ -7,17 +7,22 @@ ENV DEBIAN_FRONTEND=noninteractive \
+    libvirt-daemon-system \
+    libvirt-clients \
     locales \
+    openjdk-8-jre-headless \
```
Q: does the SDK require JDK 8? Can we use the generic/versionless package to be more future-proof?

jgraham replied:
That would be nice, wouldn't it. Unfortuantely until I used this older version I was getting some mysterious failures, so I think it's really required.

Hexcles reviewed (approved):
LGTM
